From a702791cdce64923c2fde0583241c8f73a3ae055 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Mon, 5 Sep 2016 17:41:25 +0300
Subject: [PATCH 0470/1345] fix: mvpp2x: forward 802.3X FC packets to CPU

commit  7d97fa1e6c1230b0c2e9d91b844e115444b98fae from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- Issue: MAC errors in RX descriptor
- 802.3X forwarded to CPU and do not create MAC errors
- fix Jira SYSTEMSW-2889

Change-Id: Ie7510f85ff277411121158911930217e5c7728ee
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/32555
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c |    6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
index 75a9a34..f35ff10 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
@@ -342,6 +342,12 @@ int mv_gop110_gmac_mode_cfg(struct gop_hw *gop, struct mv_mac_data *mac)
 		return -1;
 	}
 
+	/* Control packets are forwarded into the ingress pipe */
+	val = mv_gop110_gmac_read(gop, mac_num, MV_GMAC_PORT_SERIAL_PARAM_CFG_REG);
+	U32_SET_FIELD(val, MV_GMAC_PORT_SERIAL_PARAM_CFG_FORWARD_802_3X_FC_EN_MASK,
+		      (0x1 << MV_GMAC_PORT_SERIAL_PARAM_CFG_FORWARD_802_3X_FC_EN_OFFS));
+	mv_gop110_gmac_write(gop, mac_num, MV_GMAC_PORT_SERIAL_PARAM_CFG_REG, val);
+
 	/* Jumbo frame support - 0x1400*2= 0x2800 bytes */
 	val = mv_gop110_gmac_read(gop, mac_num, MV_GMAC_PORT_CTRL0_REG);
 	U32_SET_FIELD(val, MV_GMAC_PORT_CTRL0_FRAMESIZELIMIT_MASK,
-- 
1.7.9.5

